src/klr.net45/klr.net45.cpp

```diff
@@ -8,20 +8,25 @@
 
 IKatanaManagerPtr g_katanaManager;
 
-
 extern "C" __declspec(dllexport) HRESULT __stdcall CallApplicationMain(PCALL_APPLICATION_MAIN_DATA data)
 {
     HRESULT hr = S_OK;
-    
+
     IKatanaManagerPtr manager = new ComObject<KatanaManager, IKatanaManager>();
 
     g_katanaManager = manager;
-    _HR(manager->InitializeRuntime(data->applicationBase));
-    g_katanaManager = NULL;
-
-    _HR(manager->CallApplicationMain(data->argc, data->argv));
 
-    data->exitcode = hr;
+    hr = manager->InitializeRuntime(data->applicationBase);
+    if (SUCCEEDED(hr))
+    {
+        g_katanaManager = NULL;
+        data->exitcode = manager->CallApplicationMain(data->argc, data->argv);
+    }
+    else 
+    {
+        printf_s("Failed to initalize runtime (%x)", hr);
```

Comment:
Allow me to clarify all the return values here. There are two values representing the exit codes here:

1. Return value from the method, in the form of `HRESULT`. It is always the exit code of the __runtime initialization__.
2. `data->exitcode`, in the form of `int`. When __runtime is initialized successfully__ it is the return value of __application main__ (which is defined by user code); on the other hand when __runtime initialization failed__ it is assigned the returned `HRESULT`. (I do think this part is a little bit odd, maybe we should leave the `data->exitcode` unassigned in this situation.)

So any caller should do:
1. Verify the return `HRESULT` first to ensure runtime is successfully initialized.
2. If so read the `data->exitcode` to determine the program exit status.

Unless **Helios** operate differently or I miss anything, the caller should have the chance to decipher the runtime status.

Did I missing anything?